js.go

```diff
@@ -0,0 +1,1175 @@
+		cb = func(m *Msg) { ocb(m); m.Ack() }
+	}
+
+	sub, err = js.nc.subscribe(deliver, queue, cb, ch, cb == nil)
+	if err != nil {
+		return nil, err
+	}
+	sub.jsi = &jsSub{js: js}
+
+	// If we are creating or updating let's process that request.
+	if shouldCreate {
+		// If not set default to ack explicit.
+		if cfg.AckPolicy == ackPolicyNotSet {
+			cfg.AckPolicy = AckExplicit
+		}
+		req := &JSApiCreateConsumerRequest{
+			Stream: stream,
+			Config: &cfg,
+		}
+
+		j, err := json.Marshal(req)
+		if err != nil {
+			return nil, err
+		}
+
+		var ccSubj string
+		if cfg.Durable != _EMPTY_ {
+			ccSubj = fmt.Sprintf(JSApiDurableCreateT, stream, cfg.Durable)
+		} else {
+			ccSubj = fmt.Sprintf(JSApiConsumerCreateT, stream)
+		}
+
+		resp, err := js.nc.Request(js.apiSubj(ccSubj), j, defaultRequestWait)
+		if err != nil {
+			if err == ErrNoResponders {
+				err = ErrJetStreamNotEnabled
+			}
+			sub.Unsubscribe()
+			return nil, err
+		}
+
+		var info JSApiConsumerResponse
+		err = json.Unmarshal(resp.Data, &info)
+		if err != nil {
+			sub.Unsubscribe()
+			return nil, err
+		}
+		if info.Error != nil {
+			sub.Unsubscribe()
+			return nil, errors.New(info.Error.Description)
+		}
+
+		// Hold onto these for later.
+		sub.jsi.stream = info.Stream
+		sub.jsi.consumer = info.Name
+		sub.jsi.deliver = info.Config.DeliverSubject
+	} else {
+		sub.jsi.stream = o.stream
+		sub.jsi.consumer = o.consumer
+		sub.jsi.deliver = ccfg.DeliverSubject
+	}
+
+	// If we are pull based go ahead and fire off the first request to populate.
+	if isPullMode {
+		sub.jsi.pull = o.pull
+		sub.Poll()
+	}
+
+	return sub, nil
+}
+
+func (js *js) lookupStreamBySubject(subj string) (string, error) {
+	var slr JSApiStreamNamesResponse
+	// FIXME(dlc) - prefix
+	req := &streamRequest{subj}
+	j, err := json.Marshal(req)
+	if err != nil {
+		return _EMPTY_, err
+	}
+	resp, err := js.nc.Request(js.apiSubj(JSApiStreams), j, defaultRequestWait)
+	if err != nil {
+		if err == ErrNoResponders {
+			err = ErrJetStreamNotEnabled
+		}
+		return _EMPTY_, err
+	}
+	if err := json.Unmarshal(resp.Data, &slr); err != nil {
+		return _EMPTY_, err
+	}
+	if slr.Error != nil || len(slr.Streams) != 1 {
+		return _EMPTY_, ErrNoMatchingStream
+	}
+	return slr.Streams[0], nil
+}
+
+type subOpts struct {
+	// For attaching.
+	stream, consumer string
+	// For pull based consumers, batch size for pull
+	pull int
+	// For manual ack
+	mack bool
+	// For creating or updating.
+	cfg *ConsumerConfig
+}
+
+func Durable(name string) SubOpt {
+	return func(opts *subOpts) error {
+		opts.cfg.Durable = name
+		return nil
+	}
+}
+
+func Attach(stream, consumer string) SubOpt {
+	return func(opts *subOpts) error {
+		opts.stream = stream
+		opts.consumer = consumer
+		return nil
+	}
+}
+
+func Pull(batchSize int) SubOpt {
+	return func(opts *subOpts) error {
+		if batchSize == 0 {
+			return errors.New("nats: batch size of 0 not valid")
+		}
+		opts.pull = batchSize
+		return nil
+	}
+}
+
+func ManualAck() SubOpt {
```

Comment:
if not in manual  mode what happens if I ack it anyway? I dont see special handling for that here

Reply:
Server does not care, I could try to detect and suppress in client but for now thought it was fine as is.